quinn-proto/src/connection/streams/mod.rs

```diff
@@ -249,10 +257,12 @@ impl<'a> SendStream<'a> {
     ///
     /// [`StreamEvent::Finished`]: crate::StreamEvent::Finished
     pub fn finish(&mut self) -> Result<(), FinishError> {
+        let max_send_data = self.state.max_send_data(&self.id);
         let stream = self
             .state
             .send
             .get_mut(&self.id)
+            .map(|s| s.get_or_insert_with(|| Box::new(Send::new(max_send_data))))
```

Comment:
I feel like there are a few recurring patterns here that would make good candidates for moving into a method somewhere to raise the level of abstraction. We could address that after merging this, but would be great if you can take a whack at it.